app/medusa/src/main.rs

```diff
@@ -0,0 +1,136 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+// We have to do this if we don't otherwise use it to ensure its vector table
+// gets linked in.
+extern crate stm32h7;
+
+use stm32h7::stm32h753 as device;
+
+use drv_stm32h7_startup::ClockConfig;
+
+use cortex_m_rt::entry;
+
+#[entry]
+fn main() -> ! {
+    system_init();
+
+    const CYCLES_PER_MS: u32 = 400_000;
+
+    #[cfg(feature = "traptrace")]
+    kern::profiling::configure_events_table(tracing::table());
+
+    unsafe { kern::startup::start_kernel(CYCLES_PER_MS) }
+}
+
+fn system_init() {
+    let cp = cortex_m::Peripherals::take().unwrap();
+    let p = device::Peripherals::take().unwrap();
+
+    // Check the package we've been flashed on. Medusa boards use BGA240.
+    // Gimletlet boards are very similar but use QFPs. This is designed to fail
+    // a Medusa firmware that was accidentally flashed onto a Gimletlet.
+    //
+    // We need to turn the SYSCFG block on to do this.
+    p.RCC.apb4enr.modify(|_, w| w.syscfgen().enabled());
+    cortex_m::asm::dsb();
+    // Now, we can read the appropriately-named package register to find out
+    // what package we're on.
+    match p.SYSCFG.pkgr.read().pkg().bits() {
+        0b1000 => {
+            // TFBGA240, yay
+        }
+        _ => {
+            // uh
+            panic!();
+        }
+    }
+
+    // We read the board model very early in boot to try and detect the
+    // firmware being flashed on the wrong board. In particular, we read the
+    // model _before_ setting up the clock tree below, just in case we change
+    // the crystal configuration in a subsequent rev.
+    //
+    // Note that the firmware _does not_ adapt to different board models. We
+    // still require different firmware per model; this check serves to detect
+    // if you've flashed the wrong one, only.
+    //
+    // The model is on the following pins:
+    // - ID0: PC6
+    // - ID1: PC7
+    // - ID2: PC13
+    // Un-gate the clock to GPIO bank C.
+    p.RCC.ahb4enr.modify(|_, w| w.gpiogen().set_bit());
+    cortex_m::asm::dsb();
+    // PE{13,7,6} are already inputs after reset
+    #[rustfmt::skip]
+    p.GPIOC.moder.modify(|_, w| w
+        .moder6().input()
+        .moder7().input()
+        .moder13().input());
+
+    // Unlike other designs, we aren't using any internal pullup resistors, so
+    // we won't wait for the inputs or ID traces to charge.
+
+    let id_mask = (1 << 6) | (1 << 7) | (1 << 13);
+    let model = p.GPIOC.idr.read().bits() & id_mask;
+
+    cfg_if::cfg_if! {
+        if #[cfg(target_board = "medusa-a")] {
+            let expected_model = 0b000;
+        } else {
+            compile_error!("not a recognized medusa board");
+        }
+    }
```

Comment:
at first, i wasn't super convinced that the use of `cfg_if` was really pulling its weight here, given that this could just be 
```rust
#[cfg(target_board = "medusa-a")]
let expected_model = 0b000;
#[cfg(not(target_board = "medusa-a"))]
compile_error!(...);
```

but then it occurred to me that there's the possibility of more revisions in the future, and it's probably easier to handle them with `cfg_if`, so...carry on!

Reply:
Yeah, its a bit clunky currently. But it let makes adding future support a bit less cumbersome!